service/qos/raft_service_level_distributed_data_accessor.cc

```diff
@@ -0,0 +1,105 @@
+                    nanoseconds_counter{std::chrono::duration_cast<std::chrono::nanoseconds>(d).count()}));
+        },
+    }, tv);
+}
+
+raft_service_level_distributed_data_accessor::raft_service_level_distributed_data_accessor(cql3::query_processor& qp, service::raft_group0_client& group0_client)
+    : _qp(qp)
+    , _group0_client(group0_client) {}
+
+future<qos::service_levels_info> raft_service_level_distributed_data_accessor::get_service_levels() const {
+    return qos::get_service_levels(_qp, db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2, db::consistency_level::LOCAL_ONE);
+}
+
+future<qos::service_levels_info> raft_service_level_distributed_data_accessor::get_service_level(sstring service_level_name) const {
+    return qos::get_service_level(_qp, db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2, std::move(service_level_name), db::consistency_level::LOCAL_ONE);
+}
+
+future<> raft_service_level_distributed_data_accessor::do_raft_command(std::optional<service::group0_guard> guard, std::vector<mutation> mutations, std::string_view description) const {    
+    //NOTE: abort_source here is temporary. It's fixed and passed as an argument in a later commit
+    abort_source as;
+
+    service::write_mutations change {
+        .mutations{mutations.begin(), mutations.end()},
+    };
+
+    auto group0_cmd = _group0_client.prepare_command(change, *guard, description);
+    co_await _group0_client.add_entry(std::move(group0_cmd), std::move(*guard), &as);
+}
+
+future<> raft_service_level_distributed_data_accessor::set_service_level(sstring service_level_name, qos::service_level_options slo, std::optional<service::group0_guard> guard) const {
+    if (this_shard_id() != 0) {
+        on_internal_error(logger, "raft_service_level_distributed_data_accessor: must be executed on shard 0");
+    }
+
+    if (!guard) {
+        on_internal_error(logger, "raft_service_level_distributed_data_accessor: guard must be present");
+    }
+    
+    static sstring insert_query = format("INSERT INTO {}.{} (service_level, timeout, workload_type) VALUES (?, ?, ?);", db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2);
+    data_value workload = slo.workload == qos::service_level_options::workload_type::unspecified
+            ? data_value::make_null(utf8_type)
+            : data_value(qos::service_level_options::to_string(slo.workload));
+    
+    auto timestamp = guard->write_timestamp();
+    auto muts = co_await _qp.get_mutations_internal(insert_query, qos_query_state(), timestamp, {service_level_name, timeout_to_data_value(slo.timeout), workload});
+
+    co_await do_raft_command(std::move(guard), std::move(muts), "set service level");
+}
+
+future<> raft_service_level_distributed_data_accessor::drop_service_level(sstring service_level_name, std::optional<service::group0_guard> guard) const {
+    //FIXME: remove this when `role_manager::remove_attribute()` will be done in one raft command
```

Comment:
I did it here because I wanted to avoid pulling reference to `group0_client` to `service_level_controller` and obtaining the guard here is only temporary.